### ISSUE TITLE
[OJS] CSRF token validity time (one hour) breaks the review form

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#2428]
@asmecher, I am breaking this to a new issue, because this although connected to CSRF #2417 is a bit different.

The CSRF validity time is now set to one hour, but that is, in many cases, too short. Reviewers tend to keep to form open for a long time, more than one hour. 

In a clean install of OJS3.0.2 I changed the CSRF token validity time 60 seconds (https://github.com/pkp/pkp-lib/blob/master/classes/session/Session.inc.php#L221) and opened a review form. Then waited for a bit and tried to submit the form.

The result was that a error occurred (translation strings visible etc.) and all the form data was lost. The form was also basically stuck.

While it is very likely that it takes more than an hour to fill that form, so some solution is needed here! I know that just changing the validity time is a trade off with security.

**Comment by asmecher** [comment]
@ajnyga, the attached patch (https://github.com/pkp/pkp-lib/commit/ff718bfe512b7a19653d3982b975c91cba050629) should be just cosmetic, but it'll get those untranslated keys resolved when the form is bounced back to the user. That bounce should also include the filled-in review form fields and it appears not to -- I'll take a look at that next. I realize these are tangential tweaks but it's a start; meanwhile I'm watching for the weird session invalidation issue that seems to be happening intermittently.

**Comment by ajnyga** [comment]
Thanks @asmecher ! One of our journals had a lot of trouble with this and has ended up losing at least four review reports. First I thought that it was something to do with the journal settings, but then realized that they have a longer custom review form which probably takes long time to complete.

I did a quick fix by raising the validity time so that they do not loose more reviews. 

One way to fix this, and also add a nice new feature, would be a **save button** in the form. If the form would have a save function, you could also trigger an **autosave** like in WP. The save button was actually among the new feature requests that came from our workshops.

**Comment by asmecher** [comment]
> That bounce should also include the filled-in review form fields and it appears not to -- I'll take a look at that next.

This is already the case, according to a quick test -- if the CSRF token is expired/incorrect, the user should get bounced back to the review page with a "form improperly submitted" error, but the review  form contents should be preserved, meaning they would be able to submit again without having to re-enter the review.

I suspect lost reviews are probably encountering https://github.com/pkp/pkp-lib/issues/2417 rather than CSRF expiries.

**Comment by ajnyga** [comment]
I got a new report concerning this yesterday form one of our editors. The reviewer got the "form improperly submitted" error and the contents of the review were lost.

I will do some testing with 3.1. today by setting the token validity time to 60 seconds. My first theory is that the 60 minutes is not enough when someone is writing the review straight to the form and the error is still resulting into lost data.

**Comment by ajnyga** [comment]
@asmecher the one hour validity time still leads to lost data in OJS 3.1.

I changed the validity time to 60 seconds to test this. I entered the review form, added text to the form and waited for 60 seconds. After that I submitted review.

The result was that on the top of the form you get this and all the text is lost:
![screenshot_10](https://user-images.githubusercontent.com/16347527/35767482-a49411fa-08f5-11e8-9c75-a58ef702843b.png)

Not that the actual red error window appears only when you submit the form for the second time.

I now changed the validity time in our own production server to 6 hours, but I think that this could be an issue with many journals if the reviewer happens to write the review to the form.

**Comment by asmecher** [comment]
I think that's expected behavior, @ajnyga -- if the token has expired by the time the user POSTs it, the request won't be accepted. A practical example: someone steals a CSRF token and then waits for an opportunity to use a social engineering hack to trick a user into submitting a "delete journal" request or something. If the token is old, the attack won't work.

**Comment by ajnyga** [comment]
Hi @asmecher 

This is what I originally suspected, but you did write earlier that "This is already the case, according to a quick test -- if the CSRF token is expired/incorrect, the user should get bounced back to the review page with a "form improperly submitted" error, but the review form contents should be preserved"

But as I noted above, after the "form improperly submitted" error occurs and the user returns to the form, it is empty. This is of course extremely frustrating for the reviewer.

We had a few cases like this last spring and I then changed the one hour value to 6 hours. Now after upgrading to 3.1.0.1. I forgot to do that change to the code and within a couple of weeks this happened again.

I am fairly sure that other journals are also running into this problem, but while it is hard to track down what the reviewers have done based on the reports you get from them (usually with the editor working as the middleman) this has gone unnoticed.

I think that it is fairly common for a reviewer to write a review more than an hour. Most of them of course do this outside OJS.

I can see three solutions here:

- prevent the csrf error from emptying the form (not sure if this against the whole concept of csrf)
- make the csrf validity time longer (not sure what the security issues are)
- create an autosave function to the form and trigger that for example every five minutes making the amount of lost data minimal (elsewhere I suggested that there should be a save button in the review form that let's the reviewer to come back later to finish the job).

**Comment by asmecher** [comment]
Ah, it sounds like there's a specific problem with that form, in that it's not correctly bouncing the data. No, I don't think a correctly-functioning bounce shouldn't violate the safety of a CSRF check -- it should just basically say "there was a problem" and present the form again with the submitted data pre-populated for review and submission. If it's a malicious attempt to trick a user into submitting the form, they'd have to click "submit" at the bottom of the page after seeing the "there was a problem" message.

**Comment by ajnyga** [comment]
I have a reason to believe that this same thing happened with the editorial decision form for an editor.

She contacted me and said that she had written a decision for an author but it was not saved or sent to the author.

Maybe if we consider an autosave, we could consider the editorial decision form as well.

**Comment by ajnyga** [comment]
@asmecher @NateWr I have received new reports concerning this problem. At least two reviewers have lost long review reports they have written because the token has expired.

I have even changed the expiry time to six hours, but people seem to leave the form open for a long time.

If I would go on to solve this problem, what would be the preferred solution. Is it the auto-save function?

**Comment by NateWr** [comment]
Did you ever figure out the issue with the form contents disappearing? I think the first thing would be to fix that issue, so that a subsequent resubmit will work.

But aside from that, yes, I think an auto-save function is probably the best bet.

**Comment by ajnyga** [comment]
I have to say I am a little unsure what the expected behaviour is when the CSRF token becomes unvalid. 

Should it prevent the submission of the form with javascript? Do other form preserve the content. For example in the submission form in step 3 when you enter submission metadata. If someone leaves that form open for an hour and then tries to submit it, does it really work so that the error occurs, you are bounced back to the form **and** the content is preserved?

**Comment by ajnyga** [comment]
edit: ok so it does work better in the submission form. The content is not getting lost. I will try to investigate what the difference is with the submission form and the review form that they work differently here. Suspect it is a small problem.

**Comment by ajnyga** [comment]
@NateWr 
ok, so I have been looking at this a couple of hours now, but can not find a reason why the form looses the content.

The starting point was that the error appears like this:
![35767490-d572e058-08f5-11e8-9ef1-29ce32ca5b92](https://user-images.githubusercontent.com/16347527/57077867-bcb0c700-6cf5-11e9-8eb1-d327ac5a2fbf.png)

That can be easily fixed by changing https://github.com/pkp/pkp-lib/blob/master/templates/reviewer/review/step3.tpl#L21 to 
`{include file="controllers/notification/inPlaceNotification.tpl" notificationId="reviewStep3FormNotification"}`

After that the errors start appearing like this:
<img width="1094" alt="Screen Shot 2019-05-02 at 16 27 23" src="https://user-images.githubusercontent.com/16347527/57078596-372e1680-6cf7-11e9-8fd0-60305723e5d2.png">

However, with that change the reviewer form is still loosing the content from the `comments` and `commentsPrivate` fields (and also the recommendation selection, but that is not that bad).

Both forms use the AjaxFormHandler and both forms trigger a similar `saveStep` method together with `form: readInputData` and `form:fetch` methods and create a notification call. Also the returned json in that call seems to be similar. 

The forms just respond differently with the reviewer form loosing content. Another small difference seems to be that in Submission form the notification appears as a red box in the upper right corner, but a similar inPlaceNotification in the reviewer form appears as shown above. (To make things even more complicated, I also get the other type of notification in the Submission form as well.)

It seems that I could solve this by adding this to `ReviewerReviewStep3Form:fetch`:

```
		if (strlen($comments = $this->getData('comments'))>0) {
			$templateMgr->assign('comment', $comments);
		}
		if (strlen($commentsPrivate = $this->getData('commentsPrivate'))>0) {
			$templateMgr->assign('commentPrivate', $commentsPrivate);
		}
```
here https://github.com/pkp/pkp-lib/blob/master/classes/submission/reviewer/form/ReviewerReviewStep3Form.inc.php#L85

**But** I can not see any other form needing that kind of addition? All forms I've tested (Submission, Section, Settings etc.) preserve the content in textareas even when csrf check fails. So I am pretty sure this is some kind of javascript / JSON problem that I just can not find.

I have tested this by changing the csrf value to 10 seconds so that it becomes invalid quickly: https://github.com/pkp/pkp-lib/blob/master/classes/session/Session.inc.php#L215

**Comment by NateWr** [comment]
@ajnyga :joy: I feel you. Is it possible that all the other forms use ajax submission and this one doesn't? An ajax submission would mean no page reload and so the content is "left" inside. But not sure...

**Comment by ajnyga** [comment]
thanks to Nate, this is now resolved. PR coming soon.

**Comment by ajnyga** [comment]
@NateWr @asmecher 

ojs: https://github.com/pkp/ojs/pull/2377
pkp-lib: https://github.com/pkp/pkp-lib/pull/4718

I think this is worth backporting to 3.1.2. It is a serious bug.

**Comment by asmecher** [comment]
THANKS, @ajnyga, this was a nasty typo that would've been hard to track down. I've merged and back-ported to `stable-3_1_2` for both OJS and OMP, so that this can be included in 3.1.2-1 of each.

**Comment by ajnyga** [comment]
Nate noticed that. Or to be exact I did notice it, but did not think that it would make a difference, but apparently it did :)

This was one of my oldest issues dating back to April 2017, nice to have it closed at last 🎉
